tools/chpl-language-server/src/chpl-language-server.py

```diff
@@ -664,6 +694,52 @@ async def complete(ls: ChapelLanguageServer, params: CompletionParams):
 
         return CompletionList(is_incomplete=False, items=items)
 
+    @server.feature(TEXT_DOCUMENT_CODE_ACTION)
+    async def code_action(ls: ChapelLanguageServer, params: CodeActionParams):
+        text_doc = ls.workspace.get_text_document(params.text_document.uri)
+        actions = []
+
+        prefix = "Warning: \\[Deprecation\\]:"
+        chars = "[a-zA-Z0-9_.:;\\- ]*?"
+        ident = "[a-zA-Z0-9_.]+?"
+        pat1 = f"{prefix}{chars}(?:'(?:{ident})'{chars})?'(?P<replace1>{ident})'{chars}"
+        pat2 = f"{prefix}{chars}use{chars}(?P<tick>[`' ])(?P<replace2>{ident})(?P=tick){chars}instead{chars}"
+        regex = re.compile(f"({pat1})|({pat2})")
+
+        edits_to_make: Tuple[List[Diagnostic], List[TextEdit]] = ([], [])
+
+        for d in params.context.diagnostics:
+            m = re.match(regex, d.message)
+            if m and (m.group("replace1") or m.group("replace2")):
+                original = ls.get_text(text_doc, d.range)
+                replacement = m.group("replace1") or m.group("replace2")
+                te = TextEdit(d.range, replacement)
```

Comment:
Should this pperform a text substitution rather than completely replacing the underlined text? The diagnostic could in theory highlight more than the identifier that's deprecated.